### PR TITLE
sql: support NOT VISIBLE in CREATE TABLE for user-defined hidden columns

### PR DESCRIPTION
**Pull request by Heoric**
Informs #53428 
Informs #58440

Release note (sql change): It is now possible to use the `NOT VISIBLE`
qualifier for new column definitions in CREATE TABLE. This causes the
column to become 'hidden'. Hiddens columns are not considered when
using `*` in SELECT clauses.

(Note that CockroachDB already supported hidden columns previously,
such as `rowid` which is added automatically when a table definition
has no PRIMARY KEY constraint. This change merely adds the opportunity
for end-users to define their own hidden columns.)

This feature is intended for use in combination with other features
related to geo-partitioning introduced in v21.1, to offer more control
about how geo-partitioning keys get exposed to client ORMs and other
automated tools that inspect the SQL schema.

### CODE REVIEW

**Comment by ajwerner** [comment]
This happens to be a concept we're explicitly exploring right now in this RFC: https://github.com/cockroachdb/cockroach/pull/58440

**Review by knz — COMMENTED** [review]
Thank you, this actually looks pretty good.

You will need a proper commit message however, that explains the work (motivation and approach).
Also the PR description needs to indicate the link to the motivating issue, e.g. "fixes #53428" 
As well as a release note that outlines the user-facing changes.
For more information about how to write a proper commit message, head here: https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages

Also this PR is missing unit tests. There needs to be some tests in `sql/parser/parse_test.go`, and also some tests in `pkg/sql/logictests`

I invite you to take inspiration from my previous PR #26644 which also contains unit tests (you can take most of them over). I'd, of course, also appreciate a reference to that previous work in your commit message.

> Reviewed 13 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-:-MR-JmmT3s_1Sr72VFSL:b-wdzj4n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Heoric)

___
*[pkg/sql/show_create.go, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR-JS8kDy-rO7swvA1l:-MR-JS8kDy-rO7swvA1m:b-lm2tm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63154cdfc98123dfa596d73f55d60b8f91f70a0/pkg/sql/show_create.go#L87)):*
> ```Go
> 	f.WriteString(" (")
> 	primaryKeyIsOnVisibleColumn := false
> 	// visibleCols := desc.VisibleColumns()
> ```

I don't understand this - why did you need to change it?
___
*[pkg/sql/parser/sql.y, line 1365 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR-JdNN1bFlYAn_gGtk:-MR-JdNN1bFlYAn_gGtl:b-sv6tlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63154cdfc98123dfa596d73f55d60b8f91f70a0/pkg/sql/parser/sql.y#L1365)):*
> ```Yacc
> //
> // Column qualifiers:
> //   [CONSTRAINT <constraintname>] {NULL | NOT NULL | VISIBLE | UNIQUE [WITHOUT INDEX] | PRIMARY KEY | CHECK (<expr>) | DEFAULT <expr>}
> ```

probably `| [NOT] VISIBLE ...` here
___
*[pkg/sql/parser/sql.y, line 1901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR-Jbn43RcgrXDuSser:-MR-Jbn43RcgrXDuSses:b-r8vf0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63154cdfc98123dfa596d73f55d60b8f91f70a0/pkg/sql/parser/sql.y#L1901)):*
> ```Yacc
>   }
>   // ALTER TABLE <name> ALTER [COLUMN] <colname> SET NOT VISIBLE
> | ALTER opt_column column_name SET NOT VISIBLE
> ```

please add the corresponding tests in `parse_test.go` thanks
___
*[pkg/sql/parser/sql.y, line 5655 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR-JYjtAHgRYlgrOo6p:-MR-JYjtAHgRYlgrOo6q:bmd7yhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63154cdfc98123dfa596d73f55d60b8f91f70a0/pkg/sql/parser/sql.y#L5655)):*
> ```Yacc
> //
> // Column qualifiers:
> //   [CONSTRAINT <constraintname>] {NULL | NOT NULL | VISIBLE | NOT VISIBLE |UNIQUE [WITHOUT INDEX] | PRIMARY KEY | CHECK (<expr>) | DEFAULT <expr>}
> ```

make this `| [NOT] VISIBLE | UNIQUE ...`
___
*[pkg/sql/sem/tree/create.go, line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR-Iy-1127pLssCj-pm:-MR-Iy-1127pLssCj-pn:bxw2c6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63154cdfc98123dfa596d73f55d60b8f91f70a0/pkg/sql/sem/tree/create.go#L600)):*
> ```Go
> 		ctx.FormatNode(&node.Nullable.ConstraintName)
> 	}
> 	if node.Hidden {
> ```

Can you also update `(node *ColumnTableDef) docRow()` accordingly in pretty.go, thanks.
___
*[pkg/sql/vtable/pg_catalog.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR-IpxIE7cEc_tEW9R2:-MR-IpxIE7cEc_tEW9R3:beq3q0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63154cdfc98123dfa596d73f55d60b8f91f70a0/pkg/sql/vtable/pg_catalog.go#L21)):*
> ```Go
> const PGCatalogAm = `
> CREATE TABLE pg_catalog.pg_am (
> 	oid OID NOT VISIBLE,
> ```

This change, while welcome, belongs to a different commit/PR. Please split it away from the first commit, thanks.

**Comment by knz** [comment]
@otan  @rafiss I think this should be on your radar.

**Comment by lucy-zhang** [comment]
I'd like to request that you split the `ALTER TABLE ... ALTER COLUMN` change into a separate PR to be reviewed after this one, and limit the scope of this PR to specifying hidden columns in `CREATE TABLE` and `ALTER TABLE ... ADD COLUMN` (and writing tests for those statements). Changing the visibility of columns is a more peripheral use case, and schema changes can require some care.

**Review by knz — COMMENTED** [review]
Thanks for updating the PR.

@lucy-zhang what should we do regarding ALTER? Do we ask to remove the SQL parser support entirely, or should we keep the syntax but then report an unimplemented error during planning?

> Reviewed 11 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-:-MR5KKHaBko62Q8vY7XA:bo73im0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Heoric)

___
*[pkg/sql/parser/parse_test.go, line 1452 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR5Joqg76iil-VEgqcT:-MR5Joqg76iil-VEgqcU:b-k9hys1) ([raw file](https://github.com/cockroachdb/cockroach/blob/098c1f925eb9bb148adec0c5e142e97f7e746338/pkg/sql/parser/parse_test.go#L1452)):*
> ```Go
> 		{`ALTER TABLE a ALTER COLUMN b DROP NOT NULL`},
> 		{`ALTER TABLE a ALTER COLUMN b DROP STORED`},
> 		{`ALTER TABLE a ALTER COLUMN b SET VISIBLE`},
> ```

I think you can remove the ALTER tests here now.
___
*[pkg/sql/parser/sql.y, line 1346 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR5JyMy9pLtpns4ElzD:-MR5JyMy9pLtpns4ElzE:bwddl3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/098c1f925eb9bb148adec0c5e142e97f7e746338/pkg/sql/parser/sql.y#L1346)):*
> ```Yacc
> //   ALTER TABLE ... ALTER [COLUMN] <colname> DROP NOT NULL
> //   ALTER TABLE ... ALTER [COLUMN] <colname> DROP STORED
> //   ALTER TABLE ... ALTER [COLUMN] <colname> SET [NOT] VISIBLE
> ```

Since you are extracting the ALTER functionality into a different PR, you can remove this from this PR too.
___
*[pkg/sql/parser/sql.y, line 1901 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR5K2c48keSZoYfPtzR:-MR5K2c48keSZoYfPtzS:b-5h08c7) ([raw file](https://github.com/cockroachdb/cockroach/blob/098c1f925eb9bb148adec0c5e142e97f7e746338/pkg/sql/parser/sql.y#L1901)):*
> ```Yacc
>   }
>   // ALTER TABLE <name> ALTER [COLUMN] <colname> SET NOT VISIBLE
> | ALTER opt_column column_name SET NOT VISIBLE
> ```

same here.
___
*[pkg/sql/sem/tree/alter_table.go, line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MR5KBI0-lh5HV_yczt3:-MR5KBI0-lh5HV_yczt4:b-2guby3) ([raw file](https://github.com/cockroachdb/cockroach/blob/098c1f925eb9bb148adec0c5e142e97f7e746338/pkg/sql/sem/tree/alter_table.go#L76)):*
> ```Go
> func (*AlterTableSetAudit) alterTableCmd()           {}
> func (*AlterTableSetDefault) alterTableCmd()         {}
> func (*AlterTableSetHidden) alterTableCmd()          {}
> ```

same here

**Comment by Heoric** [comment]
Should I just commit the `create table... `and split the `ALTER TABLE ... ALTER COLUMN` syntax and implementation to another PR? @lucy-zhang @knz

**Review by knz — APPROVED** [review]
I think this looks good now.
Lucy do you want to take a look and merge this if you're satisfied?

> Reviewed 3 of 5 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-:-MR_7r1Q6u9dnKZlSvep:b-h9zkrq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Heoric)

**Comment by knz** [comment]
@Heoric sorry I forgot - you need to populate a commit message that explains the change, according to the instructions in the wiki (see my link above). Also you can squash the 3 commits together.

**Comment by Heoric** [comment]
Thank you for your advice and look forward to your review. @lucy-zhang

**Comment by Heoric** [comment]
Thank you very much for the review!




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "kena"<notifications@github.com&gt;; 
发送时间: 2021年1月21日(星期四) 晚上9:30
收件人: "cockroachdb/cockroach"<cockroach@noreply.github.com&gt;; 
抄送: "奔跑"<312747101@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [cockroachdb/cockroach]  [WIP] user-defined hidden columns (#58923)





 
@knz approved this pull request.
 
I think this looks good now.
 Lucy do you want to take a look and merge this if you're satisfied?
  
Reviewed 3 of 5 files at r3, 2 of 2 files at r4.
 Reviewable status:  complete! 0 of 0 LGTMs obtained (waiting on @Heoric)

**Review by thoszhang — COMMENTED** [review]
> Reviewed 4 of 13 files at r1, 1 of 11 files at r2, 2 of 5 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58923)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Heoric)

___
*[pkg/sql/logictest/testdata/logic_test/hidden_columns, line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MRxHVWR828mks3Y1qmd:-MRxHVWR828mks3Y1qme:b-ux4lo9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0535075d42cd8b9b59544e7c1fe291669d064d96/pkg/sql/logictest/testdata/logic_test/hidden_columns#L1)):*
> ```
> statement ok
> ```

Can you add logic tests for these interactions with DDL statements:

- dropping a hidden column
- renaming a hidden column
- adding an index on a hidden column
- adding a foreign key constraint on a hidden column
- adding a foreign key constraint that references a hidden column
- `ALTER PRIMARY KEY` on a primary key with hidden columns

I expect these to all work, but it'll be good to make sure.

**Review by thoszhang — COMMENTED** [review]
> Reviewed 1 of 5 files at r3, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58923)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Heoric)

___
*[pkg/sql/logictest/testdata/logic_test/hidden_columns, line 70 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MS5YSmNAwdUOlYWnys8:-MS5YSmO7FscLIFrmM39:ba06saz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c508bce541cef2db0fa1d7f16ff3a2904675ca20/pkg/sql/logictest/testdata/logic_test/hidden_columns#L70)):*
> ```
> 
> statement ok
> ALTER TABLE kv Drop COLUMN x;
> ```

nit: capitalize `DROP`
___
*[pkg/sql/logictest/testdata/logic_test/hidden_columns, line 75 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MS5YpUUEiDRIwDdwt4K:-MS5YpUUEiDRIwDdwt4L:b-9qvkh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c508bce541cef2db0fa1d7f16ff3a2904675ca20/pkg/sql/logictest/testdata/logic_test/hidden_columns#L75)):*
> ```
> SELECT x FROM kv;
> ```

nit: Only 1 newline between these cases
___
*[pkg/sql/logictest/testdata/logic_test/hidden_columns, line 82 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MS5YuOPEqTf3KOQUP6N:-MS5YuOPEqTf3KOQUP6O:b-wbrk2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c508bce541cef2db0fa1d7f16ff3a2904675ca20/pkg/sql/logictest/testdata/logic_test/hidden_columns#L82)):*
> ```
> 
> statement ok
> CREATE TABLE t2(b INT NOT VISIBLE, c INT, d INT, PRIMARY KEY (d), FOREIGN KEY(b) REFERENCES t1(b));
> ```

Can you add some queries after all these schema changes to help make sure that the new descriptor states are usable?
___
*[pkg/sql/logictest/testdata/logic_test/hidden_columns, line 103 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-MS5YWC5DZWSW3r9shA6:-MS5YWC5DZWSW3r9shA7:bmw87u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c508bce541cef2db0fa1d7f16ff3a2904675ca20/pkg/sql/logictest/testdata/logic_test/hidden_columns#L103)):*
> ```
> 
> ```

nit: Extra newlines at the end (the linter doesn't pick this up)

**Comment by Heoric** [comment]
This has a error when I add a foreign key constraint that references a hidden column(including rowid)。
Is this reasonable?
Should I change it or keep the original error report？ @lucy-zhang @knz

**Comment by knz** [comment]
@lucy-zhang this error surprises me too. What do you think?

**Comment by lucy-zhang** [comment]
That error is coming from executing the internal query generated here:

https://github.com/cockroachdb/cockroach/blob/3e2a4e7f8891d153b4785a6a337522481c0ffedb/pkg/sql/check.go#L152

I reproduced this on master and got the query from the logs:

```
root@:26257/defaultdb> CREATE TABLE t4(c INT);
CREATE TABLE
root@:26257/defaultdb> CREATE TABLE t5(a INT);
CREATE TABLE
root@:26257/defaultdb> ALTER TABLE t4 ADD FOREIGN KEY (c) REFERENCES t5 (rowid);
ERROR: validate fk constraint: column "t.rowid" does not exist
SQLSTATE: 42703
```

```
I210129 15:59:27.481167 3052 sql/check.go:290 ⋮ [n1,job=628705089054212097,scExec,id=52,mutation=1] 118  validating FK ‹"fk_c_ref_t5"› (‹"t4"› [‹[c rowid]›] -> ‹"t5"› [‹[rowid]›]) with query ‹"SELECT s.c, s.rowid FROM \n\t\t  (SELECT c, rowid FROM [52 AS src]@{IGNORE_FOREIGN_KEYS} WHERE c IS NOT NULL) AS s\n\t\t\tLEFT OUTER JOIN\n\t\t\t(SELECT * FROM [53 AS target]) AS t\n\t\t\tON s.c = t.rowid\n\t\t WHERE t.rowid IS NULL  LIMIT 1"›
```

Here it is cleaned up a bit:

```
SELECT s.c, s.rowid
  FROM (SELECT c, rowid FROM [52 AS src]@{IGNORE_FOREIGN_KEYS} WHERE c IS NOT NULL) AS s
       LEFT JOIN (SELECT * FROM [53 AS target]) AS t ON s.c = t.rowid
 WHERE t.rowid IS NULL
 LIMIT 1;
```

This query produces the same error when issued from the sql console, so it's not a problem with the internal executor or anything. I'll ask around and probably file a separate issue, but I don't think that should block this PR.

**Review by thoszhang — APPROVED** [review]
Thanks, changes LGTM. Could you please squash your commits?

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-:-MSTIZLp9Oldpae4GwKF:b-pgglk0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Heoric)

**Review by knz — COMMENTED** [review]
Ok this is looking good now, but the commit message is imperfect.
@heoric I will take care of editing the commit message for you. Thank you for your change!

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58923#-:-MSbliHJ-ZTO89GCmrVk:b-6sxalt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Heoric)

**Comment by knz** [comment]
bors r=lucy-zhang,knz

**Comment by kurokochin** [comment]
@otan thanks for covering me! sorry, this week is quite a hectic week for me as I become the on-call for my team, hopefully, in my next contribution I'll work from the start till the end.

**Pull request merged**
